### PR TITLE
Extract Nexus API to scippnexus

### PR DESCRIPTION
**Pull request by SimonHeybrock**
This creates a scippnexus "package". Note that it does not split the packing yet. I think we will do that when we split this off into its own repo.

### CODE REVIEW

File: src/scippneutron/file_loading/_json_nexus.py [hunk 1]
```diff
@@ -3,9 +3,9 @@
 # @author Matthew Jones
 from __future__ import annotations
 from typing import Tuple, Dict, List, Any, Union
-import numpy as np
-from ._common import Group, MissingAttribute
 from dataclasses import dataclass
+import numpy as np
+from scippnexus._common import Group, MissingAttribute
```
Comment (jl-wynen) [inline src/scippneutron/file_loading/_json_nexus.py:8]: Should move those into a public module.

File: src/scippnexus/nxevent_data.py [hunk 2]
```diff
@@ -132,12 +131,6 @@ def _getitem(self, select: ScippIndex) -> sc.DataArray:
         return sc.DataArray(data=binned, coords={'event_time_zero': event_time_zero})
 
     def _check_for_missing_fields(self):
-        if contains_stream(self._group):
-            # Do not warn about missing datasets if the group contains
-            # a stream, as this will provide the missing data
-            raise SkipSource("Data source is missing datasets"
-                             "but contains a stream source for the data")
```
Comment (jl-wynen) [inline src/scippnexus/nxevent_data.py:139]: Do we no longer need this?

File: src/scippnexus/nxobject.py [hunk 3]
```diff
@@ -72,8 +76,7 @@ def __contains__(self, name: str) -> bool:
     def __getitem__(self, name: str) -> Any:
         attr = self._attrs[name]
         # Is this check for string attributes sufficient? Is there a better way?
-        is_json = isinstance(self._attrs, JSONAttributeManager)
-        if isinstance(attr, (str, bytes)) and not is_json:
```
Comment (jl-wynen) [inline src/scippnexus/nxobject.py:76]: What happens in the case of json?

File: src/scippnexus/nxobject.py [hunk 4]
```diff
@@ -138,8 +141,7 @@ def attrs(self) -> Attrs:
     @property
     def dtype(self) -> str:
         dtype = self._dataset.dtype
-        if str(dtype).startswith('str') or (not isinstance(self._dataset, JSONDataset)
-                                            and h5py.check_string_dtype(dtype)):
+        if str(dtype).startswith('str') or h5py.check_string_dtype(dtype):
```
Comment (jl-wynen) [inline src/scippnexus/nxobject.py:144]: Hard coded dependency on h5py. Can this be refactored to decouple it from the loader?

File: src/scippnexus/nxevent_data.py (hunk 2 above)
Comment (SimonHeybrock) [reply src/scippnexus/nxevent_data.py:139]: I think we do, but I do not understand the workaround, or how this works. I don't know how we can resolve this on our own though, probably need input from ECDC.

File: src/scippnexus/nxobject.py (hunk 3 above)
Comment (SimonHeybrock) [reply src/scippnexus/nxobject.py:76]: As far as I can tell (from the tests) it is always utf8 (and this is how I implemented the workaround), so nothing will be done.

File: src/scippnexus/nxobject.py (hunk 4 above)
Comment (SimonHeybrock) [reply src/scippnexus/nxobject.py:144]: Are you suggesting to re-implement `check_string_dtype`?

File: src/scippneutron/file_loading/_json_nexus.py (hunk 1 above)
Comment (SimonHeybrock) [reply src/scippneutron/file_loading/_json_nexus.py]: Added `scippnexus.typing`.

File: src/scippnexus/nxobject.py (hunk 4 above)
Comment (SimonHeybrock) [reply src/scippnexus/nxobject.py:144]: What problem would we solve though?

**Pull request merged**